pandas/core/frame.py

```diff
@@ -2305,6 +2305,35 @@ def is_dtype_instance_mapper(column, dtype):
         dtype_indexer = include_these & exclude_these
         return self.loc[com._get_info_slice(self, dtype_indexer)]
 
+    def astype(self, dtype, copy=True, raise_on_error=True, **kwargs):
+        """
+        Cast object to given data type(s).
+        Return a copy when copy = True (be really careful with this!)
+
+        Parameters
+        ----------
+        dtype : numpy.dtype or Python type (to cast entire DataFrame to the same
+                type). Alternatively, {col: dtype}, where col is a column index
+                or column label and dtype is a numpy.dtype or Python type (to
+                cast one or more of the DataFrame's columns to column-specific
+                types).
+        raise_on_error : raise on invalid input
+        kwargs : keyword arguments to pass on to the constructor
+
+        Returns
+        -------
+        casted : type of caller
+        """
+        if isinstance(dtype, collections.Mapping):
+            df = DataFrame(data=self, copy=copy, **kwargs)
+            for col_key, col_dtype in dtype.items():
+                df.ix[:, col_key] = df.ix[:, col_key].astype(col_dtype)
```

Comment:
this iterative updating will be quite inefficient as it will potentially copy the entire new dtype each time as it grows; this should change with the internals refactor #11970 , but for now do something like:

```
astyped_cols = pd.concat([ df[col_key].astype(col_dtype) for col_key, col_dtype in dtype.items() ], axis=1)
return pd.concat([astyped_cols, df[[df.columns.difference(dtype.keys())]], axis=1).reindex(columns=df.columns)
```

which should minimize the copying.